File: java-to-go-projects/13-event-emitter/Main.java
package events;

/**
 * Demo: Event emitter with various listener patterns.
 *
 * Go structure:
 *   main.go                  — Demo usage
 *   events/listener.go       — Listener struct { ID, Callback func(interface{}), Once bool }
 *   events/emitter.go        — Emitter struct with sync.RWMutex
 *                               On/Once/Off/Emit/ListenerCount/EventNames
 *                               map[string][]Listener for event→listeners
 *                               Wildcard "*" support
 *
 * Rust structure:
 *   main.rs                  — Demo usage
 *   listener.rs              — Listener<T> with Box<dyn Fn(&T) + Send + Sync>
 *   emitter.rs               — Emitter<T> with HashMap<String, Vec<Listener<T>>>
 *                               Behind Arc<RwLock<>> for thread safety
 *                               on/once/off/emit/listener_count/event_names
 *
 * Key learning:
 *   - Go: func type as callback, closures, sync.RWMutex, slice manipulation
 *   - Rust: Box<dyn Fn> for stored callbacks, trait bounds (Send+Sync),
 *           lifetime issues with closures, Arc<RwLock<>>
 */
public class Main {

    public static void main(String[] args) {
        Emitter<String> emitter = new Emitter<>();

        // Persistent listener
        String id1 = emitter.on("user:login", data ->
                System.out.println("Login event: " + data));

        // One-shot listener
        emitter.once("user:login", data ->
                System.out.println("First login only: " + data));

        // Wildcard listener — catches all events
        emitter.on("*", data ->
                System.out.println("  [wildcard] got: " + data));

        // Another event
        emitter.on("user:logout", data ->
                System.out.println("Logout event: " + data));

        System.out.println("Events: " + emitter.eventNames());
        System.out.println("Login listeners: " + emitter.listenerCount("user:login"));

        System.out.println("\n--- Emit user:login (alice) ---");
        emitter.emit("user:login", "alice");

        System.out.println("\n--- Emit user:login (bob) ---");
        emitter.emit("user:login", "bob");
        // Note: "First login only" won't fire the second time

        System.out.println("\n--- Emit user:logout (alice) ---");
        emitter.emit("user:logout", "alice");

        System.out.println("\n--- Remove login listener and emit ---");
        emitter.off("user:login", id1);
        emitter.emit("user:login", "charlie");
        // Only wildcard fires

        System.out.println("\nLogin listeners: " + emitter.listenerCount("user:login"));
    }
}

File: java-to-go-projects/13-event-emitter/main.go
package eventemitter
package main

import (
	"fmt"
	"sync"
	"sync/atomic"
)

type Listener struct {
	ID          int
	Callback    func(interface{})
	Once        bool
	InvokeCount atomic.Int64
}

// invoke calls the listener's callback. Returns true if it was a one-shot listener.
func (l *Listener) invoke(data interface{}) bool {
	panic("todo")
}

type Emitter struct {
	listeners map[string][]*Listener
	nextID    int
	mu        sync.RWMutex
}

func newEmitter() *Emitter {
	panic("todo")
}

// on registers a persistent listener. Returns listener ID.
func (e *Emitter) on(event string, callback func(interface{})) int {
	panic("todo")
}

// once registers a one-shot listener. Returns listener ID.
func (e *Emitter) once(event string, callback func(interface{})) int {
	panic("todo")
}

// off removes a specific listener by event name and ID. Returns true if found.
func (e *Emitter) off(event string, id int) bool {
	panic("todo")
}

// offAll removes all listeners for an event.
func (e *Emitter) offAll(event string) {
	panic("todo")
}

// emit fires an event. Invokes all matching listeners + wildcard "*" listeners.
// Removes one-shot listeners after invocation.
func (e *Emitter) emit(event string, data interface{}) {
	panic("todo")
}

// listenerCount returns the number of listeners for a specific event.
func (e *Emitter) listenerCount(event string) int {
	panic("todo")
}

// eventNames returns all event names that have listeners.
func (e *Emitter) eventNames() []string {
	panic("todo")
}

func main() {
	fmt.Println("Event Emitter Demo")
	panic("todo: register persistent, one-shot, and wildcard listeners; emit events; remove listeners; print summary")
}
